Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/AWSCredentialIdentityResolvers/IMDSAWSCredentialIdentityResolver.swift

```diff
@@ -6,19 +6,202 @@
+                apiVersion = .extended
+            }
+        } catch IMDSError.metadata(.nonRetryable(let message)) where message.contains("404") {
+            if apiVersion == .unknown {
+                // Fallback to .legacy and recurse.
+                apiVersion = .legacy
+                return try await getIdentity(identityProperties: identityProperties)
+            } else if ec2InstanceProfileName == nil {
+                // Remove profile name previously resolved from IMDS and recurse.
+                resolvedProfile = nil
+                return try await getIdentity(identityProperties: identityProperties)
+            } else {
+                throw IMDSError.invalidProfileName
+            }
+        } catch {
+            throw IMDSError.invalidProfileName
+        }
+
+        return try parseJSONCredentials(jsonCredentialString: jsonCredentialString)
+    }
+
+    private func throwIfIMDSDisabled() throws {
+        let fileBasedConfig = try CRTFileBasedConfiguration()
+        let imdsDisabled = FieldResolver(
+            configValue: nil,
+            envVarName: "AWS_EC2_METADATA_DISABLED",
+            configFieldName: "disable_ec2_metadata",
+            fileBasedConfig: fileBasedConfig,
+            converter: { Bool($0.lowercased()) }
+        ).value ?? false
+        guard !imdsDisabled else {
+            throw AWSCredentialIdentityResolverError.failedToResolveAWSCredentials(
+                "IMDSAWSCredentialIdentityResolver: "
+                + "IMDS is disabled via env var or shared config file."
             )
+        }
+    }
+
+    private func resolveCredentialURLBase() -> String {
+        return apiVersion == .legacy
+        ? "/latest/meta-data/iam/security-credentials/"
+        : "/latest/meta-data/iam/security-credentials-extended/"
+    }
+
+    private func parseJSONCredentials(
+        jsonCredentialString: String
+    ) throws -> AWSCredentialIdentity {
+        guard let jsonCredentialData = jsonCredentialString.data(using: .utf8) else {
+            throw IMDSError.deserializationError(
+                "IMDSAWSCredentialIdentityResolver: "
+                + "Failed to decode JSON credential response from IMDS."
+            )
+        }
+        let jsonCredentials = try JSONDecoder().decode(
+            JSONCredentialResponse.self,
+            from: jsonCredentialData
+        )
+        return AWSCredentialIdentity(
+            accessKey: jsonCredentials.accessKeyID,
+            secret: jsonCredentials.secretAccessKey,
+            accountID: jsonCredentials.accountID,
+            expiration: jsonCredentials.expiration,
+            sessionToken: jsonCredentials.token
         )
     }
 }
+
+private struct JSONCredentialResponse: Codable {
+    let code: String
+    let lastUpdated: Date
+    let type: String
+    let accessKeyID: String
+    let secretAccessKey: String
+    let token: String
+    let expiration: Date
+    let accountID: String?
```

Comment:
All of the above fields will definitely be present (except `accountID`)?  Because parsing this type will fail and throw if any one of them are not present

Reply:
Yep.